colcon_cmake/task/cmake/build.py

```diff
@@ -206,6 +216,40 @@ def _store_cmake_args(self, build_base, cmake_args):
     def _get_last_cmake_args_path(self, build_base):
         return Path(build_base) / 'cmake_args.last'
 
+    def _have_jobs_arg(self, cmd, env):
+        """
+        Check the given command line argument list looking for jobs args.
+
+        There are a number of ways to specify the number of jobs to use. This
+        function checks if any known mechanisms are present. This checks for
+        the following patterns: [-j#, -j #, --parallel, '-- -j']. The last item
+        is for pre CMake 3.12, while the prior options are for 3.12+ where
+        CMake knows how to pass jobs to the make system.
+
+        :param list cmd: Command line argument list for the build command.
+        :returns: True when a jobs argument is found in cmd.
+        """
+        have_dashes = False
+        cmake_has_jobs = False
+        cmake_ver = get_cmake_version()
+        if cmake_ver and cmake_ver >= parse_version('3.12.0'):
+            cmake_has_jobs = True
+            # Respect CMAKE_BUILD_PARALLEL_LEVEL environment variable.
+            cmake_parallel_level = env.get('CMAKE_BUILD_PARALLEL_LEVEL', '')
+            if cmake_parallel_level:
+                return True
+        for arg in cmd:
+            if arg == '--':
+                # Now passing through args to the make system
+                have_dashes = True
+            if re.match('-j[0-9]*', arg) and (cmake_has_jobs or have_dashes):
```

Comment:
This regex is insufficient to catch other job-like flags in Make.

Reply:
With respect to the functional overlap, I agree there's some conceptual overlap between `_have_job_arg` and `_get_make_arguments`, but the logic is distict. They are both capturing information from different places and for different use cases. `_have_job_arg` is testing whether the command line has specified a number of jobs to use, while `_get_make_arguments` is implicitly looking up `MAKEFLAGS` from the environment variable for similar arguments in order to pass this through to the build. Really `_have_job_arg` is trying to ascertain whether we need to call `_get_make_arguments`.

I'm not sure how this could gracefully merged into `_get_make_arguments` becasue we don't actually need to do anything when `_have_job_arg` is true. The semantics of the name `_get_make_arguments` really don't fit this either. The name implies it's pulling various `make` arguments into the command line being constructed, but it actually only extracts the job arguments.

Overall I'm not a fan of the `MAKEFLAGS` extraction primarily because it's taking an environment variable associated with `make` and implicitly using it with other, non `make` programs. Presumbly, invoking `make` will already get the job arguments from `MAKEFLAGS` because that's what `MAKEFLAGS` is designed to do. Ninja doesn't look like it has an equivalent and this looks like a bit of a hack and explicitly supporting job arguments when invoking colcon makes it obsolute.

I'd argue that coclon is actually subverting `MAKEFLAGS` and would suggest that having this PR makes this `MAKEFLAGS` usage obsolete. Both functions could be deprecated and removed in future.

I had concidered renaming `_get_make_arguments` but decided to keep it siloed and avoid diving to deeply into the intended behaviour of `_get_make_arguments`.

Reply:
With respect to the regular expression what flags are you thinking it will miss?

Reply:
> With respect to the regular expression what flags are you thinking it will miss?

All of these are flags affect the job server: https://github.com/colcon/colcon-cmake/blob/ecf417d03b67c4ad4734f042cb73c42e61a70376/colcon_cmake/task/cmake/build.py#L294-L302

Reply:
I was actually avoiding handling `-l` becuause that's on its own doesn't use jobs.

However, your link does highlight a way I could resolve both issues above. Would you be satisfied with pulling out the job args parsing regular expressions into its own function for both `_have_job_arg ` and `_get_make_arguments` to use? This way the parsing is unified, but each function can continue to parse from different places as is needed.

Reply:
Actually, some of these are not valid for the command line and are only valid for `make`. It looks like `ninja` doesn't only uses the `-j` and `-l` abbreviations and does not support `--jobs` or `--load-average`. I can see the complexity of determining whether to parse `MAKEFLAGS` or not exploding.